java/src/main/native/src/row_conversion.cu

```diff
@@ -373,140 +1079,647 @@ static int calc_fixed_width_kernel_dims(const cudf::size_type num_columns,
+                          std::vector<row_batch> const &row_batches) {
+  auto const desired_rows_and_columns = static_cast<int>(sqrt(shmem_limit_per_tile));
+  auto const column_stride = util::round_up_unsafe(
+      [&]() {
+        if (desired_rows_and_columns > num_columns) {
+          // not many columns, group it into 8s and ship it off
+          return std::min(8, num_columns);
+        } else {
+          return util::round_down_safe(desired_rows_and_columns, 8);
+        }
+      }(),
+      JCUDF_ROW_ALIGNMENT);
+
+  // we fit as much as we can given the column stride
+  // note that an element in the table takes just 1 bit, but a row with a single
+  // element still takes 8 bytes!
+  auto const bytes_per_row =
+      util::round_up_safe(util::div_rounding_up_unsafe(column_stride, 8), JCUDF_ROW_ALIGNMENT);
+  auto const row_stride =
+      std::min(num_rows, util::round_down_safe(shmem_limit_per_tile / bytes_per_row, 64));
+
+  std::vector<detail::tile_info> validity_tile_infos;
+  for (int col = 0; col < num_columns; col += column_stride) {
+    int current_tile_row_batch = 0;
+    int rows_left_in_batch = row_batches[current_tile_row_batch].row_count;
+    int row = 0;
+    while (row < num_rows) {
+      if (rows_left_in_batch == 0) {
+        current_tile_row_batch++;
+        rows_left_in_batch = row_batches[current_tile_row_batch].row_count;
+      }
+      int const tile_height = std::min(row_stride, rows_left_in_batch);
+
+      validity_tile_infos.emplace_back(detail::tile_info{
+          col, row, std::min(col + column_stride - 1, num_columns - 1), row + tile_height - 1});
+      row += tile_height;
+      rows_left_in_batch -= tile_height;
+    }
+  }
+
+  return validity_tile_infos;
+}
+
+template <typename RowSize> struct row_size_functor {
+  RowSize _row_sizes;
+  size_type _num_rows;
+  row_size_functor(RowSize row_sizes) : _row_sizes(row_sizes){};
+
+  CUDA_DEVICE_CALLABLE
+  uint64_t operator()(int row_index) { return static_cast<uint64_t>(_row_sizes[row_index]); }
+};
+
+/**
+ * @brief Builds batches of rows that will fit in the size limit of a column.
+ *
+ * @tparam RowSize iterator that gives the size of a specific row of the table.
+ * @param num_rows Total number of rows in the table
+ * @param row_sizes iterator that gives the size of a specific row of the table.
+ * @param all_fixed_width bool indicating all data in this table is fixed width
+ * @param stream stream to operate on for this work
+ * @param mr memory resource used to allocate any returned data
+ * @returns vector of size_type's that indicate row numbers for batch boundaries and a
+ * device_uvector of row offsets
+ */
+
+template <typename RowSize>
+batch_data build_batches(size_type num_rows, RowSize row_sizes, bool all_fixed_width,
+                         rmm::cuda_stream_view stream, rmm::mr::device_memory_resource *mr) {
+  auto uint64_row_sizes =
+      cudf::detail::make_counting_transform_iterator(0, row_size_functor(row_sizes));
```

Comment:
The call site passes a `counting_transform_iterator` for `row_sizes`. (I have remarked that this could probably be replaced with a `constant_iterator`.)
In addition, I suspect that `total_size` can be computed without further wrapping `row_sizes` in another `counting_transform_iterator`:
```c++
auto const total_size = thrust::reduce(rmm::exec_policy(stream), row_sizes, row_sizes + num_rows);
```

It appears that `row_size_functor()` does cast the result from `size_type` to `uint64_t`. If the `row_sizes` iterator returned `uint64_t`, would that not help obviate the counting iterator, and `row_size_functor`?

Reply:
I'll try looking into the build error that results from making this change.